### ISSUE TITLE
catastrophic cluster failure

### ISSUE THREAD

**Issue by temujin9** [flynn/flynn#2610]
Not certain the original cause, but initial complaint from monitoring was loss of a scheduler. https://gist.github.com/dbf8464b1c8661edc735 https://gist.github.com/b2a3888d77da2d1b1a44 https://gist.github.com/a82293750eefefb1b537

**Comment by titanous** [comment]
Discoverd remediation is #2612.

@lmars will follow-up with analysis of the scheduler bug that was triggered by discoverd registrations flapping due to very slow disk IO.

**Comment by lmars** [comment]
During the discoverd flaps, the scheduler decided that the discoverd formation (along with some other formations) was in the wrong state:


From https://gist.github.com/dbf8464b1c8661edc735:

```
t=2016-03-21T02:42:32+0000 lvl=info msg="expected different from actual" component=scheduler fn=formationDiff app.id=275d0efd-6b77-4d76-95e7-bc59b814466f release.id=69bab3d7-636c-4352-b85f-c8a6e539fb6b expected=map[app:0] actual=map[app:3] diff=map[app:-3]
t=2016-03-21T02:42:32+0000 lvl=info msg="formation in incorrect state" component=scheduler fn=handleFormationDiff app.id=275d0efd-6b77-4d76-95e7-bc59b814466f release.id=69bab3d7-636c-4352-b85f-c8a6e539fb6b diff=map[app:-3]
t=2016-03-21T02:42:32+0000 lvl=info msg="stopping 3 app jobs" component=scheduler fn=handleFormationDiff app.id=275d0efd-6b77-4d76-95e7-bc59b814466f release.id=69bab3d7-636c-4352-b85f-c8a6e539fb6b
```

I haven't got to the bottom of why this was the case, but I am thinking the scheduler should do an extra check to make sure there is another active formation when scaling down a formation for a system app.

@titanous WDYT?

**Comment by titanous** [comment]
Yeah, that makes sense but we'll need to mark critical and non-critical apps for that purpose.

**Comment by titanous** [comment]
The discoverd flaps were fixed in #2618.

**Comment by temujin9** [comment]
Okay, I think this is now addressed. Closing it; I'll reopen if I see a recurrence.
